### PR TITLE
support dpdk

### PR DESCRIPTION
**Pull request by luoyunhe**
#### What type of this PR
Feature

run Kube-OVN with nodes which run ovs-dpdk or ovs-kernel

### CODE REVIEW

**Comment by oilbeater** [comment]
@luoyunhe the  `tmp/Dockerfile` should be removed

**Comment by luoyunhe** [comment]
> @luoyunhe the `tmp/Dockerfile` should be removed

ok, Could you give me some suggestions for this PR, thanks.

**Comment by oilbeater** [comment]
@luoyunhe it looks good to me now. I will test the pr in my cluster later and will let you know if something need to change

File: pkg/daemon/handler.go [hunk 1]
```diff
@@ -242,6 +258,68 @@ func (csh cniServerHandler) handleAdd(req *restful.Request, resp *restful.Respon
...
+		return fmt.Errorf("can not fount volume %s in pod %s", volumeName, pod.Name)
+	}
+	if volume.EmptyDir == nil {
+		return fmt.Errorf("volume %s is not empty dir", volume.Name)
+	}
+	originSharedDir := fmt.Sprintf("/var/lib/kubelet/pods/%s/volumes/kubernetes.io~empty-dir/%s", pod.UID, volumeName)
+	newSharedDir := getShortSharedDir(pod.UID, volumeName)
+	_, err = os.Stat(newSharedDir)
+	if os.IsNotExist(err) {
+		err = os.MkdirAll(newSharedDir, 0750)
```
Comment (oilbeater) [inline pkg/daemon/handler.go:279]: error handle format

File: pkg/daemon/handler.go [hunk 2]
```diff
@@ -242,6 +258,68 @@ func (csh cniServerHandler) handleAdd(req *restful.Request, resp *restful.Respon
...
+		return nil
+
+	}
+	return err
+
+}
+
+func removeShortSharedDir(pod *v1.Pod, volumeName string) (err error) {
+	sharedDir := getShortSharedDir(pod.UID, volumeName)
+	_, err = os.Stat(sharedDir)
```
Comment (oilbeater) [inline pkg/daemon/handler.go:300]: error handle format

File: pkg/daemon/handler.go [hunk 3]
```diff
@@ -348,6 +426,17 @@ func (csh cniServerHandler) handleDel(req *restful.Request, resp *restful.Respon
 		var nicType string
 		if podRequest.DeviceID != "" {
 			nicType = util.OffloadType
+		} else if podRequest.VhostUserSocketVolumeName != "" {
+			nicType = util.DpdkType
+			err = removeShortSharedDir(pod, podRequest.VhostUserSocketVolumeName)
```
Comment (oilbeater) [inline pkg/daemon/handler.go:431]: error handle format

File: pkg/daemon/handler.go [hunk 4]
```diff
@@ -242,6 +258,68 @@ func (csh cniServerHandler) handleAdd(req *restful.Request, resp *restful.Respon
...
+
+	}
+	return err
+
+}
+
+func removeShortSharedDir(pod *v1.Pod, volumeName string) (err error) {
+	sharedDir := getShortSharedDir(pod.UID, volumeName)
+	_, err = os.Stat(sharedDir)
+	if os.IsNotExist(err) {
```
Comment (oilbeater) [inline pkg/daemon/handler.go:301]: other error should also be reported

**Review by oilbeater — CHANGES_REQUESTED** [review]

File: pkg/daemon/config.go [hunk 5]
```diff
@@ -144,6 +144,11 @@ func (config *Configuration) initNicConfig(nicBridgeMappings map[string]string)
 		klog.Infof("Find node tunnel interface name: %v", nodeTunnelName)
 	}
 
+	isDPDKNode := node.GetLabels()[util.OvsDpTypeLabel] == "userspace"
+
+	if isDPDKNode {
+		config.Iface = "br-phy"
```
[the remark this answers is not in the document]
Comment (luoyunhe) [reply pkg/daemon/config.go:150]: 这里获取的IP会作为隧道封包IP，使用这个IP可以使用户态的DPDK直接收到封包数据，关于用户态的隧道配置可以参照这个OvS的文档：https://github.com/openvswitch/ovs/blob/master/Documentation/howto/userspace-tunneling.rst

File: pkg/daemon/handler.go [hunk 6]
```diff
@@ -242,6 +258,68 @@ func (csh cniServerHandler) handleAdd(req *restful.Request, resp *restful.Respon
...
+func createShortSharedDir(pod *v1.Pod, volumeName string) (err error) {
+	var volume *v1.Volume
+	for index, v := range pod.Spec.Volumes {
+		if v.Name == volumeName {
+			volume = &pod.Spec.Volumes[index]
+			break
+		}
+	}
+	if volume == nil {
+		return fmt.Errorf("can not fount volume %s in pod %s", volumeName, pod.Name)
```
[the remark this answers is not in the document]
Comment (luoyunhe) [reply pkg/daemon/handler.go:270]: fix

File: docs/dpdk-hybrid.md [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+```bash
+# specify encap IP
+ENCAP_IP=192.168.122.193/24
+# specify pci device
+DPDK_DEV=0000:00:0b.0
+```
+
+
+## Set up Kube-OVN
+Just run install.sh --with-hybrid-dpdk
```
[the remark this answers is not in the document]
Comment (luoyunhe) [reply docs/dpdk-hybrid.md:31]: ok

File: pkg/daemon/config.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (oilbeater) [reply pkg/daemon/config.go:150]: 这个名字应该是可配的，不应该写死在这里

**Review by oilbeater — CHANGES_REQUESTED** [review]

File: pkg/daemon/config.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (luoyunhe) [reply pkg/daemon/config.go]: 已经完成修改

**Review by oilbeater — APPROVED** [review]

**Comment by oilbeater** [comment]
Great work! Thanks @luoyunhe

**Pull request merged**